Enhance Challenge Tag Filtering with "Has Coding Challenge" Option

bogminic opened the pull request:
### Description
This pull request adds a new filter option to the Challenge Tags dropdown called "Has Coding Challenge". When selected, it filters and displays only the challenges that include a Coding Challenge section, helping users easily identify and focus on coding-related tasks.

### Motivation
As part of my involvement in promoting security awareness, I’ve delivered a few hacking presentations, and my colleagues particularly enjoyed the **Coding Challenges** in Juice Shop. This new filter aims to make those challenges easier to find and encourage more people to explore them.

### Demo
![has-coding-challange-2](https://github.com/user-attachments/assets/134e47d7-372e-48af-ab02-16e19fc44a85)

### Affirmation

- [x] My code follows the [CONTRIBUTING.md](https://github.com/juice-shop/juice-shop/blob/master/CONTRIBUTING.md) guidelines

J12934 commented on frontend/src/app/score-board/score-board.component.ts:
```diff
@@ -154,6 +161,7 @@ export class ScoreBoardComponent implements OnInit, OnDestroy {
   }
 
   filterAndUpdateChallenges (): void {
+    console.log(this.allChallenges)
```
```suggestion
```

J12934 reviewed (changes requested):
Hi 👋

cool, nice to hear that the coding challenges are working nicely for your use case :)

I think filtering them with tags is fine.
I would prefer if we'd move the tag creation to the backend though. This way the tag is actually part of the challenge json, so it's present for everybody and its not just a frontend thing.

The place to add it would be the data creator https://github.com/juice-shop/juice-shop/blob/master/data/datacreator.ts#L64, to check it there we might also want to add a `hasCodingChallenge` function to the coding challenge lib. https://github.com/juice-shop/juice-shop/blob/master/lib/codingChallenges.ts

bogminic commented:
> Hi 👋
> 
> cool, nice to hear that the coding challenges are working nicely for your use case :)
> 
> I think filtering them with tags is fine. I would prefer if we'd move the tag creation to the backend though. This way the tag is actually part of the challenge json, so it's present for everybody and its not just a frontend thing.
> 
> The place to add it would be the data creator https://github.com/juice-shop/juice-shop/blob/master/data/datacreator.ts#L64, to check it there we might also want to add a `hasCodingChallenge` function to the coding challenge lib. https://github.com/juice-shop/juice-shop/blob/master/lib/codingChallenges.ts

Hi Jannik,

Thanks for the feedback! I agree that moving the tag creation to the backend makes sense. I'll update the data creator (`datacreator.ts`) and add the `hasCodingChallenge` function to the coding challenges library (`codingChallenges.ts`) as suggested.

I’ll make these changes when I have some spare time and update the PR accordingly.

Best regards,
Bogdan

bkimminich commented:
I agree that we should make this appear as an actual tag, just one that is dynamically added and not pre-defined in the `challenges.yml`. The tag name could just be "With Coding Challenge" or "Has Coding Challenge". We should then consider in a second step to rename the "Code Analysis" tag because those will be mixed up by people for sure :

![image](https://github.com/user-attachments/assets/eac83ca5-6f81-4d2e-a09c-94cf093afe06)

bogminic commented:
> I agree that we should make this appear as an actual tag, just one that is dynamically added and not pre-defined in the `challenges.yml`. The tag name could just be "With Coding Challenge" or "Has Coding Challenge". We should then consider in a second step to rename the "Code Analysis" tag because those will be mixed up by people for sure :
> 
> ![image](https://private-user-images.githubusercontent.com/3531020/437373938-eac83ca5-6f81-4d2e-a09c-94cf093afe06.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU1OTA2MzcsIm5iZiI6MTc0NTU5MDMzNywicGF0aCI6Ii8zNTMxMDIwLzQzNzM3MzkzOC1lYWM4M2NhNS02ZjgxLTRkMmUtYTA5Yy05NGNmMDkzYWZlMDYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDQyNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA0MjVUMTQxMjE3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9M2FmMjQzZWUzY2YxYjM0MjIxMWFmNDk5NWQ1YjY5ODFmM2RlOTgwYzdjNzU2MmVjYmE1Mjk4ZGE0N2ZjMGNlZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.ExBoO6af2z5n9chHgDs0SBhoVH2QvIKih-caJ73QUQE)

Thanks! I agree that "**With Coding Challenge**" is a good name for the tag—happy to go with that.

Also, totally makes sense to avoid confusion with the existing "**Code Analysis**" tag. We could rename it to "**Code Check**" or "**Code Dive**". What do you think about these two options?

bogminic commented:
Hi everyone,

I’ve moved the logic for `hasCodingChallenge` and the addition of the new tag ("With Coding Challenge") to the backend. I also updated the i18n files and cleaned up unnecessary frontend code (removed the call to `CodeSnippetService.challenges` and one related test).

We could also remove the `CodeSnippetService.challenges` method (and its test), as it’s no longer used in the frontend codebase. However, I’m not sure if it might be needed in the future.

What do you think — should I go ahead and remove `CodeSnippetService.challenges`?

J12934 commented:
Will have another look and try :)

If it's unused definitely best to remove, if we need it again we can go back to it in git.

J12934 commented:
Nice, cool that this has also removed one request from the score-board to the server 🙌

J12934 commented:
quickly removed the service and the endpoint in the backend :)
